### PR TITLE
Added options to specify maximum number of consecutive failures

### PR DESCRIPTION
**Pull request by kdima**
Fix for #8

### CODE REVIEW

File: grpc/naming.go [hunk 1]
```diff
@@ -84,10 +100,15 @@ func (w *watcher) Next() ([]*naming.Update, error) {
...
+	for {
 		freshTargets, err := w.resolver.Lookup(w.domainName)
 		if err != nil {
 			lastErr = err
+			consecutiveErrors += 1
+			if consecutiveErrors >= w.maxConsecutiveErrors && w.maxConsecutiveErrors != 0 {
+				break
+			}
 			continue
 		}
```
Comment (improbable-yifan) [inline grpc/naming.go]: should reset the consecutiveErr count on successful lookup

File: grpc/naming.go [hunk 2]
```diff
@@ -84,10 +100,15 @@ func (w *watcher) Next() ([]*naming.Update, error) {
...
 
 	var lastErr error
-	for i := 0; i < MaximumConsecutiveErrors; i++ {
+	consecutiveErrors := 0
+	for {
 		freshTargets, err := w.resolver.Lookup(w.domainName)
 		if err != nil {
 			lastErr = err
+			consecutiveErrors += 1
+			if consecutiveErrors >= w.maxConsecutiveErrors && w.maxConsecutiveErrors != 0 {
```
Comment (improbable-yifan) [inline grpc/naming.go]: do you not need a special case for detecting the `-1` for `macConsecutiveErrors`?

File: grpc/naming.go [hunk 3]
```diff
@@ -19,25 +19,39 @@ import (
...
+func WithMaximumConsecutiveErrors(maxErr int) SrvResolverOptions {
+	return func (resolver *resolver) {
+		resolver.maxConsecutiveErrors = maxErr
+	}
 }
 
 // New creates a gRPC naming.Resolver that is backed by an SRV lookup resolver.
-func New(srvResolver srv.Resolver) naming.Resolver {
-	return &resolver{srvResolver: srvResolver}
+func New(srvResolver srv.Resolver, options ...SrvResolverOptions) naming.Resolver {
```
Comment (improbable-yifan) [inline grpc/naming.go:47]: given that there's only one option, perhaps make it a required parameter which removes the need for the Option funcs

File: grpc/naming.go [hunk 4]
```diff
@@ -19,25 +19,39 @@ import (
...
+	res :=  &resolver{srvResolver: srvResolver, maxConsecutiveErrors: -1}
```
Comment (improbable-yifan) [inline grpc/naming.go]: could extract out a value for `NoRetryLimit = -1`

**Review by improbable-yifan — CHANGES_REQUESTED** [review]

File: grpc/naming.go (hunk 3 above)
Comment (kdima) [reply grpc/naming.go:47]: Want to keep the external api unchaged. Otherwise this will be a breaking change for users.

File: grpc/naming.go (hunk 1 above)
Comment (kdima) [reply grpc/naming.go]: we do not need to as after the first successful lookup we will exit the loop

**Comment by kdima** [comment]
@yifanzz updated ptal

File: grpc/naming.go [hunk 5]
```diff
@@ -84,10 +103,15 @@ func (w *watcher) Next() ([]*naming.Update, error) {
...
 
 	var lastErr error
-	for i := 0; i < MaximumConsecutiveErrors; i++ {
+	consecutiveErrors := 0
+	for {
 		freshTargets, err := w.resolver.Lookup(w.domainName)
 		if err != nil {
 			lastErr = err
+			consecutiveErrors += 1
+			if consecutiveErrors >= w.maxConsecutiveErrors && w.maxConsecutiveErrors != NoRetryLimit {
```
Comment (improbable-yifan) [inline grpc/naming.go]: This line might read nicer if you swap the two conditions as we want to make clear that if NoRetryLimit is specified then nothing else is worth comparing.

```
w.maxConsecutiveErrors != NoRetryLimit && onsecutiveErrors >= w.maxConsecutiveErrors
```

**Review by improbable-yifan — APPROVED** [review]

**Pull request merged**